icechunk/src/repository.rs

```diff
@@ -725,6 +734,63 @@ impl Repository {
         }
     }
 
+    pub async fn rebase(
+        &mut self,
+        solver: &dyn ConflictSolver,
+        update_branch_name: &str,
+    ) -> RepositoryResult<()> {
+        let ref_data =
+            fetch_branch_tip(self.storage.as_ref(), update_branch_name).await?;
+
+        if ref_data.snapshot == self.snapshot_id {
+            // nothing to do, commit should work without rebasing
+            Ok(())
+        } else {
+            let current_snapshot =
+                self.storage.fetch_snapshot(&ref_data.snapshot).await?;
+            // FIXME: this should be the whole ancestry not local
+            let anc = current_snapshot.local_ancestry().map(|meta| meta.id);
+            let new_commits = iter::once(ref_data.snapshot.clone())
+                .chain(anc.take_while(|snap_id| snap_id != &self.snapshot_id))
+                .collect::<Vec<_>>();
+
+            // TODO: this clone is expensive
+            // we currently need it to be able to process commits one by one without modifying the
+            // changeset in case of failure
+            // let mut changeset = self.change_set.clone();
+
+            // we need to reverse the iterator to process them in order of oldest first
+            for snap_id in new_commits.into_iter().rev() {
+                let tx_log = self.storage.fetch_transaction_log(&snap_id).await?;
```

Comment:
Great question. That's what I had initially, and it's what sounded more intuitive to me too. Following is why I gave up on it on this first pass, we can always go back eventually:
* The logic o merging the transaction logs is not trivial. In fact, it looks a lot like the conflict resolution logic. Example: if one commit deletes an array and the next commit creates a group on the same path, and edits the attributes, etc. It gets very messy, and it will be pretty crazy once we introduce `rename`
* In the current design I was very careful in that if you are able to rebase only half of the commits, you get back a repo  rebased to that point, you can change your policy (or do some manual changes) and continue with another rebase. If we initially merge all the tx logs it's all or nothing.
* Memory usage is much larger if you have to merge many tx logs.
* Error reporting becomes untractable under tx log merge

I should have written a design document with all these details, but I was to bored of this problem already hehe.

As i mentioned somewhere, I don't feel I got to the best design for this problem. I'm pretty sure we'll have to revisit soon.